### ISSUE TITLE
Primary key should imply unique key in PostgreSQL

### ISSUE THREAD

**Issue by k-bx** [yesodweb/persistent#1075]
Currently, you can't do `upsert` without having an additional unique key.

```
share
  [mkPersist sqlSettings, mkMigrate "migrateAll"]
  [persistLowerCase|
-- Maps entity to its proper form
NamedPropers
    entity T.Text
    proper T.Text
    Primary entity
    UniqueEntity entity  -- <- doesn't work without this
|]

P.upsert v [NamedPropersProper P.=. T.toLower name]
```

It would be nice to not have to hold an additional index just for that, since having a primary key implies it to be unique.

**Comment by parsonsmatt** [comment]
This should be fixed by the designs in #1037 - can you review some of that discussion and see how you like it?

**Comment by k-bx** [comment]
@parsonsmatt wow, nice! Thanks for all the work. I've subscribed to that ticket, as soon as you have some sort of MVP I'd be happy to test it out on my projects to check what problems there are in practice if any.
